File: e2e/page-objects/TodoPage.js
// Page Object Model for TODO App
class TodoPage {
  constructor(page) {
    this.page = page;
    
    // Selectors
    this.taskNameInput = '#todoName';
    this.statusSelect = '#todoStatus';
    this.priorityRadio = (value) => `input[name="priority"][value="${value}"]`;
    this.addButton = '#addTodoBtn';
    this.todoList = '#todoList';
    this.todoItem = '.todo-item';
    this.deleteButton = (taskName) => `//div[contains(@class, 'todo-item')]//div[contains(text(), '${taskName}')]/following::button[contains(text(), 'Delete')]`;
    this.editControls = '.edit-controls';
    this.saveChangesButton = (taskName) => `//div[contains(@class, 'todo-item')]//div[contains(text(), '${taskName}')]/ancestor::div[contains(@class, 'todo-item')]//button[contains(text(), 'Save Changes')]`;
    this.emptyState = 'p';
  }

  async goto() {
    await this.page.goto('/');
    await this.page.waitForLoadState('networkidle');
  }

  async addTask(name, priority = '1', status = 'not started') {
    await this.page.fill(this.taskNameInput, name);
    await this.page.check(this.priorityRadio(priority));
    await this.page.selectOption(this.statusSelect, status);
    await this.page.click(this.addButton);
    await this.page.waitForTimeout(500); // Wait for API call
  }

  async getTaskByName(name) {
    return this.page.locator(`.todo-item:has-text("${name}")`).first();
  }

  async getTaskCount() {
    return await this.page.locator(this.todoItem).count();
  }

  async clickTask(name) {
    const task = await this.getTaskByName(name);
    await task.click();
  }

  async deleteTask(name, confirm = true) {
    const deleteBtn = this.page.locator(this.deleteButton(name)).first();
    await deleteBtn.click();
    
    this.page.on('dialog', async dialog => {
      if (confirm) {
        await dialog.accept();
      } else {
        await dialog.dismiss();
      }
    });
    
    await this.page.waitForTimeout(500);
  }

  async editTask(name, newPriority = null, newStatus = null) {
    await this.clickTask(name);
    await this.page.waitForTimeout(300); // Wait for edit controls to appear
    
    const task = await this.getTaskByName(name);
    
    if (newPriority) {
      const priorityRadio = task.locator(`input[name^="edit-priority"][value="${newPriority}"]`);
      await priorityRadio.check({ force: true }); // Force check even if not visible
    }
    
    if (newStatus) {
      const statusSelect = task.locator('select[id^="edit-status"]');
      await statusSelect.selectOption(newStatus);
    }
    
    const saveButton = task.locator('button:has-text("Save Changes")');
    await saveButton.click({ force: true });
    await this.page.waitForTimeout(500);
  }

  async getTaskPriority(name) {
    const task = await this.getTaskByName(name);
    const priorityText = await task.locator('.priority-badge').textContent();
    return priorityText.trim();
  }

  async getTaskStatus(name) {
    const task = await this.getTaskByName(name);
    const statusText = await task.locator('.status-badge').textContent();
    return statusText.trim();
  }

  async getTaskHighlightColor(name) {
    const task = await this.getTaskByName(name);
    const className = await task.getAttribute('class');
    
    if (className.includes('status-highlight-not-started')) return 'red';
    if (className.includes('status-highlight-in-progress')) return 'orange';
    if (className.includes('status-highlight-completed')) return 'green';
    return 'none';
  }

  async isEditControlsVisible(name) {
    const task = await this.getTaskByName(name);
    const editControls = task.locator(this.editControls);
    return await editControls.isVisible();
  }

  async isTaskInEditMode(name) {
    const task = await this.getTaskByName(name);
    const className = await task.getAttribute('class');
    return className.includes('editing');
  }

  async getEmptyStateText() {
    const emptyMsg = this.page.locator('#todoList p, #todoList');
    if (await emptyMsg.count() > 0) {
      return await emptyMsg.first().textContent();
    }
    return '';
  }

  async clearAllTasks() {
    const count = await this.getTaskCount();
    for (let i = 0; i < count; i++) {
      const firstTask = this.page.locator(this.todoItem).first();
      const deleteBtn = firstTask.locator('button:has-text("Delete")');
      
      this.page.once('dialog', dialog => dialog.accept());
      await deleteBtn.click();
      await this.page.waitForTimeout(300);
    }
  }

  async getConsoleMessages() {
    const messages = [];
    this.page.on('console', msg => messages.push(msg.text()));
    return messages;
  }

  async pressEnter() {
    await this.page.keyboard.press('Enter');
  }

  async getAlertMessage() {
    return new Promise((resolve) => {
      this.page.once('dialog', dialog => {
        resolve(dialog.message());
        dialog.accept();
      });
    });
  }
}

module.exports = { TodoPage };

File: tests/pages/todo_page.py
"""Page Object Model for TODO List Application"""
from playwright.sync_api import Page, expect
from typing import Literal


class TodoPage:
    """Page object for the TODO List application"""
    
    def __init__(self, page: Page, base_url: str = "http://localhost:3000"):
        self.page = page
        self.base_url = base_url
        
        # Locators
        self.task_name_input = page.locator("#todoName")
        self.add_button = page.locator("#addTodoBtn")
        self.status_select = page.locator("#todoStatus")
        self.todo_list = page.locator("#todoList")
        self.todo_items = page.locator(".todo-item")
        self.empty_message = page.locator(".empty-state")
        
    def navigate(self):
        """Navigate to the TODO app"""
        self.page.goto(self.base_url)
        self.page.wait_for_load_state("networkidle")
        
    def get_priority_radio(self, priority: str):
        """Get priority radio button by value"""
        return self.page.locator(f'input[name="priority"][value="{priority}"]')
    
    def add_task(self, name: str, priority: str = "1", status: str = "not started"):
        """Add a new task"""
        self.task_name_input.fill(name)
        self.get_priority_radio(priority).check()
        self.status_select.select_option(status)
        self.add_button.click()
        expect(self.get_task_by_name(name)).to_be_visible(timeout=3000)
        
    def get_task_by_name(self, name: str):
        """Get task element by name"""
        return self.page.locator(f'.todo-item:has-text("{name}")')
    
    def get_task_priority(self, task_name: str) -> str:
        """Get task priority text"""
        task = self.get_task_by_name(task_name)
        priority_text = task.locator(".priority-badge").text_content()
        return priority_text.strip()
    
    def get_task_status(self, task_name: str) -> str:
        """Get task status text"""
        task = self.get_task_by_name(task_name)
        status_text = task.locator(".status-badge").text_content()
        return status_text.strip()
    
    def get_task_highlight_class(self, task_name: str) -> str:
        """Get the highlight class of a task"""
        task = self.get_task_by_name(task_name)
        class_attr = task.get_attribute("class")
        
        if "status-highlight-not-started" in class_attr:
            return "red"
        elif "status-highlight-in-progress" in class_attr:
            return "orange"
        elif "status-highlight-completed" in class_attr:
            return "green"
        return "none"
    
    def click_task(self, task_name: str):
        """Click on a task to toggle edit mode"""
        task = self.get_task_by_name(task_name)
        task.click()
        
    def is_task_in_edit_mode(self, task_name: str) -> bool:
        """Check if task is in edit mode"""
        task = self.get_task_by_name(task_name)
        class_attr = task.get_attribute("class")
        return "editing" in class_attr
    
    def edit_task_priority(self, task_name: str, new_priority: str):
        """Change task priority in edit mode"""
        task = self.get_task_by_name(task_name)
        task_id = task.get_attribute("data-id")
        priority_radio = self.page.locator(f'input[name="edit-priority-{task_id}"][value="{new_priority}"]')
        priority_radio.check()
        
    def edit_task_status(self, task_name: str, new_status: str):
        """Change task status in edit mode"""
        task = self.get_task_by_name(task_name)
        task_id = task.get_attribute("data-id")
        status_select = self.page.locator(f'#edit-status-{task_id}')
        status_select.select_option(new_status)
        
    def save_task_changes(self, task_name: str):
        """Click Save Changes button for a task"""
        task = self.get_task_by_name(task_name)
        save_button = task.locator('button:has-text("Save Changes")')
        save_button.click()
        
    def delete_task(self, task_name: str, confirm: bool = True):
        """Delete a task"""
        task = self.get_task_by_name(task_name)
        delete_button = task.locator('button:has-text("Delete")')
        
        # Set up dialog handler before clicking
        if confirm:
            self.page.once("dialog", lambda dialog: dialog.accept())
        else:
            self.page.once("dialog", lambda dialog: dialog.dismiss())
        
        delete_button.click()
        if confirm:
            expect(self.get_task_by_name(task_name)).to_have_count(0, timeout=3000)
        
    def get_task_count(self) -> int:
        """Get the number of tasks in the list"""
        return self.todo_items.count()
    
    def is_empty_state_visible(self) -> bool:
        """Check if empty state message is visible"""
        return self.empty_message.is_visible()
    
    def get_empty_state_text(self) -> str:
        """Get empty state message text"""
        return self.empty_message.text_content().strip()
    
    def task_exists(self, task_name: str) -> bool:
        """Check if a task exists in the list"""
        return self.get_task_by_name(task_name).count() > 0
    
    def press_enter_on_input(self):
        """Press Enter key on the task name input"""
        self.task_name_input.press("Enter")
        
    def get_alert_text(self) -> str:
        """Get alert dialog text (needs to be set up with dialog handler)"""
        # This should be used with page.on("dialog") handler
        pass
    
    def is_task_name_input_focused(self) -> bool:
        """Check if task name input has focus"""
        return self.task_name_input.evaluate("el => el === document.activeElement")
    
    def get_task_name_input_value(self) -> str:
        """Get the current value of task name input"""
        return self.task_name_input.input_value()
    
    def get_console_messages(self) -> list:
        """Get console messages (needs to be set up with console listener)"""
        pass
    
    def wait_for_api_detection(self):
        """Wait for API detection to complete"""
        self.page.wait_for_timeout(1000)  # Give time for API detection
